components/ProposalEditor.tsx

```diff
@@ -0,0 +1,346 @@
+import { isValidAddress } from 'util/isValidAddress'
+import { ProposalMessageType } from '../models/proposal/messageMap'
+import { EmptyProposal, Proposal } from '../models/proposal/proposal'
+import {
+  ProposalAction,
+  ProposalRemoveMessage,
+} from '../models/proposal/proposalActions'
+import { ProposalReducer } from '../models/proposal/proposalReducer'
+import {
+  getActiveMessageId,
+  getMessage,
+  proposalMessages,
+  topmostId,
+} from '../models/proposal/proposalSelectors'
+import { labelForMessage, makeSpendMessage } from '../util/messagehelpers'
+import LineAlert from './LineAlert'
+import { MessageSelector } from './MessageSelector'
+import SpendEditor from './SpendEditor'
+
+let _editedJSON: any = undefined
+
+export function ProposalEditor({
+  initialProposal,
+  loading,
+  error,
+  onProposal,
+  contractAddress,
+  recipientAddress,
+}: {
+  initialProposal?: Proposal
+  loading?: boolean
+  error?: string
+  onProposal: (proposal: Proposal) => void
+  contractAddress: string
+  recipientAddress: string
+}) {
+  const [proposal, dispatch] = useReducer(ProposalReducer, {
+    ...(initialProposal || EmptyProposal),
+  })
+  const [editingJson, setEditingJson] = useState<boolean>(false)
+  let activeId = getActiveMessageId(proposal)
+  const activeMessage = activeId ? getMessage(proposal, activeId) : undefined
+
+  const messageActions = [
+    {
+      label: 'Spend',
+      id: 'spend',
+      action: () => addMessage(ProposalMessageType.Spend),
+      href: '#',
+      enabled: () => true,
+    },
+    {
+      label: 'Custom',
+      id: 'custom',
+      action: () => addMessage(ProposalMessageType.Custom),
+      href: '#',
+      enabled: () => true,
+    },
+    {
+      label: 'Mint',
+      id: 'mint',
+      action: () => addMessage(ProposalMessageType.Mint),
+      href: '#',
+      enabled: () => false,
+    },
+    {
+      label: 'Text',
+      id: 'text',
+      action: () => addMessage(ProposalMessageType.Text),
+      href: '#',
+      enabled: () => false,
+    },
+    {
+      label: 'Collect',
+      id: 'collect',
+      action: () => addMessage(ProposalMessageType.Collect),
+      href: '#',
+      enabled: () => false,
+    },
+  ]
+
+  let modeEditor = null
+
+  const handleCustomMessageChange = (json: any) => {
+    let action: ProposalAction | undefined
+    if (activeId) {
+      action = {
+        type: 'updateMessage',
+        id: activeId,
+        message: json,
+        valid: true,
+      }
+    } else if (activeMessage) {
+      action = {
+        type: 'addMessage',
+        messageType: ProposalMessageType.Custom,
+        message: json,
+        valid: true,
+      }
+    }
+    if (action) {
+      dispatch(action)
+    }
+  }
+
+  switch (activeMessage?.messageType) {
+    case ProposalMessageType.Spend:
+      let amount = ''
+      if (activeMessage?.message) {
+        amount = (activeMessage.message as any).bank?.send?.amount[0]?.amount
+      }
+      modeEditor = (
```

Comment:
I broke out the raw json editor into its own component. I think ProposalEditor is the thing that understands its own current mode. Otherwise "ModeEditor" is just a fancy wrapper for a switch statement.

Reply:
i'm not against `ModeEditor` just being a fancy wrapper for a switch statement 😉 

i'm cool with you leaving it here for now like this as we iterate 👍  as complexity grows we may want to consider a component